blog: modeling strings with cpp bindings

RobDickinson opened the pull request:
I struggled to figure this out in working on the KV prototype, so figured this would make a good post.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/pmem.github.io/55)
<!-- Reviewable:end -->

pbalcer commented on _posts/2017-01-23-cpp-strings.md:
```diff
@@ -0,0 +1,150 @@
+when modifying a newly allocated object, as the entire memory range of a new
+object is already in the current transaction.
+
+So for best performance, we could add `init_key` and `init_value` setter
```
While this paragraph is technically true, the performance benefit is *very* slight, on the border of negligible. This is not a good advice to give because the cost (complicated implementation) overweights the benefit.

tomaszkapela commented on _posts/2017-01-23-cpp-strings.md:
```diff
@@ -0,0 +1,150 @@
+
+#### Using fixed-length strings inside persistent structs
+
+While `p<char[size]>` is the obvious starting point for modeling fixed-size
```
`p<>` does not implement the subscript `operator[]` and therefore p<char[]> will not work, as there is no `operator=` for arrays. This is one of those not apparent problems. We have thought about make such constructs work, but `p<>` has a simple API and we would like to keep it this way and not do template magic do push in some more functionality.

tomaszkapela commented on _posts/2017-01-23-cpp-strings.md:
```diff
@@ -0,0 +1,150 @@
+};
+
+struct MyAssociations {
+  p<MyAssociation> preferred_association;
```
The problem I see here is that there is no way to overload the `operator.` Maybe you meant `persistent_ptr<MyAssociation>`?

tomaszkapela commented on _posts/2017-01-23-cpp-strings.md:
```diff
@@ -0,0 +1,150 @@
+So for best performance, we could add `init_key` and `init_value` setter
+methods, which don't call `pmemobj_tx_add_range_direct` by convention, and so
+would only be safe to use in the context of a newly allocated object. These
+would be slightly faster than `set_key` and `set_value`.
```
We have discussed this at length today with @pbalcer and @marcinslusarz, and although this is perfectly valid code, adding a new API for this slight improvement is probably not the best idea. It will complicate the client code with little performance gain.

tomaszkapela commented on _posts/2017-01-23-cpp-strings.md:
```diff
@@ -0,0 +1,150 @@
+};
+
+struct MyAssociations {
+  p<MyAssociation> preferred_association;
```
`p<>` -> `persistent_ptr<>`?

tomaszkapela commented on _posts/2017-01-23-cpp-strings.md:
```diff
@@ -0,0 +1,150 @@
+that concept someday.
+
+*Please note that `PersistentString` above is not complete -- a finished
+implementation would at a minimum also have a method to recover
```
What kind of persistent memory recovery do you mean?

RobDickinson commented:
Thanks @tomaszkapela and @pbalcer, I revamped this today to address the points you raised. Hopefully this is moving in the right direction now. :grin:

tomaszkapela commented on _posts/2017-01-23-cpp-strings.md:
```diff
@@ -0,0 +1,155 @@
+
+#### Avoid arrays of persistent pointers to single chars
+
+Don't model a variable-length string using `persistent_ptr<char> X[]`. This
```
Now that I think about it, this should be `persistent_ptr<p<char>>`. That is because `persistent_ptr` does not track the contents of the object. Sorry about not noticing this in the previous review. One more thing to note with this approach would be to `make_persistent` each and every array element which is evil. I was also thinking about `persistent_ptr<p<char>[]>` which could actually work. It would be inefficient, you would have to assign each character separately through `operator[]` (`get_rw()` could be a workaround for this) and `operator=`, but it should actually be powerfail safe.

RobDickinson replied:
Hi @tomaszkapela,

I took a little time to really think about what we really want to emphasize here.  The points that you raise are good ones, but I'm struggling to convey all of this in a way that makes this clear.

What I settled on for this latest draft is that the main objection to the 'persistent pointers to single char' is its horrible efficiency. I removed any statements about any of these variations "working" or being "safe" and am just focusing on the efficiency angle as being horrible enough to wave off doing this.

So I know I'm leaving some out details you just suggested for this latest draft, but my thought was one justification is enough for the anti-patterns and we don't need to list every fault.  For the "avoid wrapping fixed-size arrays" case, it won't compile.  For the "avoid wrapping std::string" case, it's unsafe.  For the "avoid arrays of persistent pointers to single char" case, it's ridiculously inefficient.

Anyway, I'm curious to see what you think of this latest version.  And thanks for all the time invested in this so far!

krzycz commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmem.github.io/55)*
<!-- Sent from Reviewable.io -->

RobDickinson commented:
Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions.

---

*[_posts/2017-01-23-cpp-strings.md, line 30 at r1](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r1-30:-KbWelqIz_pRYv20r3sp:b-2bzfbv) ([raw file](https://github.com/pmem/pmem.github.io/blob/8f0f43f6bbda22fbd3ee02eab1e3181fb6ca8c28/_posts/2017-01-23-cpp-strings.md#L30)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

The `p<>` template has 0 size overhead. It's size is exactly the same as the template argument's. Maybe you meant `persistent_ptr<char> []`? That would be highly inefficient and would have all the properties you mentioned. Using `p<>` has it's own downsides, each call to `operator=` is possibly a call to `pmemobj_tx_add_range_direct`, but only under certain conditions. However reading from a `p<>` is like reading any other volatile/transient field.
</blockquote></details>

Corrections made

---

*[_posts/2017-01-23-cpp-strings.md, line 37 at r1](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r1-37:-KbWei_rfWIzM-6Xsy_o:bsxizdu) ([raw file](https://github.com/pmem/pmem.github.io/blob/8f0f43f6bbda22fbd3ee02eab1e3181fb6ca8c28/_posts/2017-01-23-cpp-strings.md#L37)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

`p<>` does not implement the subscript `operator[]` and therefore p<char[]> will not work, as there is no `operator=` for arrays. This is one of those not apparent problems. We have thought about make such constructs work, but `p<>` has a simple API and we would like to keep it this way and not do template magic do push in some more functionality.
</blockquote></details>

Yes, I made this correction

---

*[_posts/2017-01-23-cpp-strings.md, line 53 at r1](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r1-53:-KbWefnhlvlhY7YBaPk1:bz9jh5y) ([raw file](https://github.com/pmem/pmem.github.io/blob/8f0f43f6bbda22fbd3ee02eab1e3181fb6ca8c28/_posts/2017-01-23-cpp-strings.md#L53)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

The problem I see here is that there is no way to overload the `operator.` Maybe you meant `persistent_ptr<MyAssociation>`?
</blockquote></details>

Yup, corrected

---

*[_posts/2017-01-23-cpp-strings.md, line 66 at r1](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r1-66:-KbWeeByrxzwX1NZgRWT:b-lg1a58) ([raw file](https://github.com/pmem/pmem.github.io/blob/8f0f43f6bbda22fbd3ee02eab1e3181fb6ca8c28/_posts/2017-01-23-cpp-strings.md#L66)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

While this paragraph is technically true, the performance benefit is *very* slight, on the border of negligible. This is not a good advice to give because the cost (complicated implementation) overweights the benefit.
</blockquote></details>

Acknowledged and removed

---

*[_posts/2017-01-23-cpp-strings.md, line 69 at r1](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r1-69:-KbWe_ZfS5neKMgz0lo0:b-lg1a58) ([raw file](https://github.com/pmem/pmem.github.io/blob/8f0f43f6bbda22fbd3ee02eab1e3181fb6ca8c28/_posts/2017-01-23-cpp-strings.md#L69)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

We have discussed this at length today with @pbalcer and @marcinslusarz, and although this is perfectly valid code, adding a new API for this slight improvement is probably not the best idea. It will complicate the client code with little performance gain.
</blockquote></details>

Acknowledged and removed

---

*[_posts/2017-01-23-cpp-strings.md, line 149 at r1](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r1-149:-KbWdw3XeJzgkzmUYvHN:b-kivvzv) ([raw file](https://github.com/pmem/pmem.github.io/blob/8f0f43f6bbda22fbd3ee02eab1e3181fb6ca8c28/_posts/2017-01-23-cpp-strings.md#L149)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

What kind of persistent memory recovery do you mean?
</blockquote></details>

I removed this comment since I'm showing a real reset() method implementation in the latest draft.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmem.github.io/55)*
<!-- Sent from Reviewable.io -->

RobDickinson commented:
Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions.

---

*[_posts/2017-01-23-cpp-strings.md, line 30 at r1](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r1-30:-KbWflYvyzRtfOAgQ-8i:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/8f0f43f6bbda22fbd3ee02eab1e3181fb6ca8c28/_posts/2017-01-23-cpp-strings.md#L30)):*
<details><summary><i>Previously, RobDickinson (Rob Dickinson) wrote…</i></summary><blockquote>

Corrections made
</blockquote></details>

Done.

---

*[_posts/2017-01-23-cpp-strings.md, line 37 at r1](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r1-37:-KbWfkaVHtYMxcafkM1U:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/8f0f43f6bbda22fbd3ee02eab1e3181fb6ca8c28/_posts/2017-01-23-cpp-strings.md#L37)):*
<details><summary><i>Previously, RobDickinson (Rob Dickinson) wrote…</i></summary><blockquote>

Yes, I made this correction
</blockquote></details>

Done.

---

*[_posts/2017-01-23-cpp-strings.md, line 53 at r1](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r1-53:-KbWfjcoXpaRvVd5mwqL:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/8f0f43f6bbda22fbd3ee02eab1e3181fb6ca8c28/_posts/2017-01-23-cpp-strings.md#L53)):*
<details><summary><i>Previously, RobDickinson (Rob Dickinson) wrote…</i></summary><blockquote>

Yup, corrected
</blockquote></details>

Done.

---

*[_posts/2017-01-23-cpp-strings.md, line 66 at r1](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r1-66:-KbWfiGkrdVRH-duCnNL:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/8f0f43f6bbda22fbd3ee02eab1e3181fb6ca8c28/_posts/2017-01-23-cpp-strings.md#L66)):*
<details><summary><i>Previously, RobDickinson (Rob Dickinson) wrote…</i></summary><blockquote>

Acknowledged and removed
</blockquote></details>

Done.

---

*[_posts/2017-01-23-cpp-strings.md, line 69 at r1](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r1-69:-KbWfh43aiEZK4XkFxNc:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/8f0f43f6bbda22fbd3ee02eab1e3181fb6ca8c28/_posts/2017-01-23-cpp-strings.md#L69)):*
<details><summary><i>Previously, RobDickinson (Rob Dickinson) wrote…</i></summary><blockquote>

Acknowledged and removed
</blockquote></details>

Done.

---

*[_posts/2017-01-23-cpp-strings.md, line 88 at r1](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r1-88:-KbWfgaQR5GbdI64AJfx:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/8f0f43f6bbda22fbd3ee02eab1e3181fb6ca8c28/_posts/2017-01-23-cpp-strings.md#L88)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

`p<>` -> `persistent_ptr<>`?
</blockquote></details>

Done.

---

*[_posts/2017-01-23-cpp-strings.md, line 149 at r1](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r1-149:-KbWffHopu8iBCgrLgKv:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/8f0f43f6bbda22fbd3ee02eab1e3181fb6ca8c28/_posts/2017-01-23-cpp-strings.md#L149)):*
<details><summary><i>Previously, RobDickinson (Rob Dickinson) wrote…</i></summary><blockquote>

I removed this comment since I'm showing a real reset() method implementation in the latest draft.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmem.github.io/55)*
<!-- Sent from Reviewable.io -->

tomaszkapela commented on _posts/2017-01-23-cpp-strings.md:
```diff
@@ -0,0 +1,156 @@
+
+Don't model a variable-length string with representations like
+`persistent_ptr<char> X[]` or `persistent_ptr<p<char>[]>`. These are
+horribly inefficient and should always be avoided. Let's consider what
```
what -> what is

tomaszkapela commented on _posts/2017-01-23-cpp-strings.md:
```diff
@@ -0,0 +1,156 @@
+Don't model a variable-length string with representations like
+`persistent_ptr<char> X[]` or `persistent_ptr<p<char>[]>`. These are
+horribly inefficient and should always be avoided. Let's consider what
+happening behind the scenes here. Each char in the variable-length string
```
Could you add that *In case of `persistent_ptr<char> X[]`, each char in the...*, because `persistent_ptr<p<char>[]>` will not have this overhead and I wouldn't want people to get confused.

tomaszkapela commented on _posts/2017-01-23-cpp-strings.md:
```diff
@@ -0,0 +1,156 @@
+horribly inefficient and should always be avoided. Let's consider what
+happening behind the scenes here. Each char in the variable-length string
+will be stored as one char in persistent memory and so each char requires
+another 16 bytes for its individual persistent pointer. So writing a
```
writing -> allocating? Writing a char does not actually modify the pointer itself.

tomaszkapela commented on _posts/2017-01-23-cpp-strings.md:
```diff
@@ -0,0 +1,156 @@
+happening behind the scenes here. Each char in the variable-length string
+will be stored as one char in persistent memory and so each char requires
+another 16 bytes for its individual persistent pointer. So writing a
+single char results in 16x more bytes written than necessary. Reading a
```
Reading -> Reading or writing ?

tomaszkapela commented on _posts/2017-01-23-cpp-strings.md:
```diff
@@ -0,0 +1,156 @@
+another 16 bytes for its individual persistent pointer. So writing a
+single char results in 16x more bytes written than necessary. Reading a
+single char results in 16x more bytes read than necessary, since every
+persistent char read requires dereferencing a 16-byte persistent pointer.
```
char read -> char access ?

tomaszkapela commented on _posts/2017-01-23-cpp-strings.md:
```diff
@@ -0,0 +1,156 @@
+single char results in 16x more bytes written than necessary. Reading a
+single char results in 16x more bytes read than necessary, since every
+persistent char read requires dereferencing a 16-byte persistent pointer.
+These are not the strings you're looking for.
```
👍

RobDickinson commented:
Review status: 0 of 1 files reviewed at latest revision, 14 unresolved discussions.

---

*[_posts/2017-01-23-cpp-strings.md, line 32 at r3](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r3-32:-KbmnB5Str9GOmepuyR8:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/5d350c6f312fb7579b32128dd671bf37de97fcd5/_posts/2017-01-23-cpp-strings.md#L32)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

what -> what is
</blockquote></details>

Done.

---

*[_posts/2017-01-23-cpp-strings.md, line 33 at r3](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r3-33:-KbmnBnk9_q3iSYqwg7E:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/5d350c6f312fb7579b32128dd671bf37de97fcd5/_posts/2017-01-23-cpp-strings.md#L33)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

Could you add that *In case of `persistent_ptr<char> X[]`, each char in the...*, because `persistent_ptr<p<char>[]>` will not have this overhead and I wouldn't want people to get confused. Well, at least it won't have the allocation overhead, the read overhead of the subscript operator will still be there.
</blockquote></details>

Done.

---

*[_posts/2017-01-23-cpp-strings.md, line 35 at r3](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r3-35:-KbmnG2929CdO_ERGMzm:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/5d350c6f312fb7579b32128dd671bf37de97fcd5/_posts/2017-01-23-cpp-strings.md#L35)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

writing -> allocating? Writing a char does not actually modify the pointer itself.
</blockquote></details>

Done.

---

*[_posts/2017-01-23-cpp-strings.md, line 36 at r3](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r3-36:-KbmnGXo6bJrzKvuiV7w:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/5d350c6f312fb7579b32128dd671bf37de97fcd5/_posts/2017-01-23-cpp-strings.md#L36)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

Reading -> Reading or writing ?
</blockquote></details>

Done.

---

*[_posts/2017-01-23-cpp-strings.md, line 38 at r3](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r3-38:-KbmnH6uc2kw_ZDF_y-K:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/5d350c6f312fb7579b32128dd671bf37de97fcd5/_posts/2017-01-23-cpp-strings.md#L38)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

char read -> char access ?
</blockquote></details>

Done.

---

*[_posts/2017-01-23-cpp-strings.md, line 39 at r3](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-KbIw3XN5HE4NJ8vr9JJ-r3-39:-KbmnHe9rpYQSUsNbn_g:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/5d350c6f312fb7579b32128dd671bf37de97fcd5/_posts/2017-01-23-cpp-strings.md#L39)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

👍 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmem.github.io/55)*
<!-- Sent from Reviewable.io -->

RobDickinson commented:
@tomaszkapela, thanks for your continued help getting this to the finish line.  I took this last round of edits as suggested without anything else. Please let me know if you spot any other problems.

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmem.github.io/55#-:-KbnIReIS4GWWmbAvosp:bnfp4nl)*
<!-- Sent from Reviewable.io -->
